features/steps/steps.py

```diff
@@ -211,14 +239,59 @@ def instance_getter(i,representation_id, representation_type, negative=False):
         if condition(i, representation_id, representation_type):
             return i
 
+def strip_split(stmt, strp = ' ', splt = ' '):
+    return list(
+        map(str.lower, map(lambda s: s.strip(strp), stmt.split(splt)))
+    )
+
+def include_subtypes(stmt):
+    stmt = strip_split(stmt, strp = '[]')
+    if len(stmt) > 1 and 'subtypes' in stmt:
+        excluding_statements = ['without', 'not', 'excluding', 'no']
+        if len(set(stmt).intersection(set(excluding_statements))):
+            return False
+        else:
+            return True
+    else:
+        return True
+
+def map_state(values, fn):
+    if isinstance(values, (tuple, list)):
+        return type(values)(map_state(v, fn) for v in values)
+    else:
+        return fn(values)
+
+@given("An {entity_opt_stmt}")
+def step_impl(context, entity_opt_stmt):
+    entity = entity_opt_stmt.split()[0]
+
+    try:
+        context.instances = context.model.by_type(entity, include_subtypes = include_subtypes(entity_opt_stmt))
+    except:
+        context.instances = []
+
+@given("{the_or_all} instances of {entity_opt_stmt}")
```

Comment:
Wait, why do we have "Given An {entity_opt_stmt}" and "Given {the_or_all} instances of {entity_opt_stmt}". Couldn't these be using the same implementation?

Reply:
Yes, I have tried that and I run into the same issue as before https://github.com/buildingSMART/ifc-gherkin-rules/issues/31 (also link to the issue so I can find it back)
It was unsure to whether to use pyparsing for this, use the brackets approach or accept a perhaps confusing distinction.

To be more precise, I thought it is easier to distinguish
`Given an IfcProduct
Given all instances of IfcGeometricRepresentations`
than
'Given the instances of IfcProduct
Given all instances of IfcGeometricRepresentations'

Reply:
In https://github.com/buildingSMART/ifc-gherkin-rules/pull/35/commits/e03228c7c4cf10dd8ad366dfdeebceab12a195df